kyokan/bob-wallet
# Start with migrate change flag

See https://github.com/handshake-org/hsd/pull/415

and especially:

https://github.com/handshake-org/hsd/pull/415#issuecomment-670677057

and

https://github.com/handshake-org/hsd/pull/415/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1e

The next release of hsd will fix a bug that corrupted the hsd wallet if the user manually generated change addresses from the API (this will affect a grand total of about 3 people worldwide, probably). The bug is already fixed but repairing corrupt wallets is an issue closed by the PR above. We had trouble deciding how much to do "automatically", the process requires a backup, database migration and possible rescan. Since the repair process is opt-in its a bit tricky to do when hsd is embedded like in Bob. This PR is a demonstration on how it should be done. I changed the hsd dependency in this PR if developers want to test.

Once this and the hsd PR are merged, the UX will be one of the following:

- User starts up Bob after upgrade, hsd checks wallet for corruption (may take a minute or so)...
  -  No corruption found, life goes on. (the change in this PR can be eventually removed, it only needs to be called once)
  - Corruption found. Because the flag is set, the wallet regenerates the missing change addresses and then **initiates a rescan**

It's a bit tricky from the `hsdMain.js` file to get the UI-blocking "wallet rescan" modal from https://github.com/kyokan/bob-wallet/pull/117, so unfortunately at this point if the user is in the second group, they may not know their wallet is rescanning and the UI may show some unexpected data. Not sure how to address this besides adding notes in the Bob changelog and release notes? 

### How to test:

- fresh data dir (dev mode)
- change the `hsd` dependency in package.json to `"hsd": "https://github.com/handshake-org/hsd#v2.1.3",`
- generate a few change addrs with cli: `$ hsw-cli --id=allison --account=default change`
- let bob sync for a while, doesn't matter. then close
- change the `hsd` dependency in package.json to `"hsd": "https://github.com/handshake-org/hsd#change-migration",`
- restart bob. 

If you remove the patch I added to `hsdMain.js`, Bob will not load completely and an error will be thrown in the console window for hsd:

```
hsdMain.js:47 Error: Wallet is corrupted.
Back up wallet and then restart with
`hsd --wallet-migrate=0` or `hs-wallet --migrate=0`
(Full node required)
    at WalletDB.migrateChange (/Users/matthewzipkin/Desktop/work/bob-wallet/node_modules/hsd/lib/wallet/walletdb.js:262)
    at async WalletDB.open (/Users/matthewzipkin/Desktop/work/bob-wallet/node_modules/hsd/lib/wallet/walletdb.js:210)
    at async Plugin.open (/Users/matthewzipkin/Desktop/work/bob-wallet/node_modules/hsd/lib/wallet/plugin.js:85)
    at async FullNode.openPlugins (/Users/matthewzipkin/Desktop/work/bob-wallet/node_modules/hsd/lib/node/node.js:448)
    at async FullNode.open (/Users/matthewzipkin/Desktop/work/bob-wallet/node_modules/hsd/lib/node/fullnode.js:257)
```

Unfortunately, the UX in Bob when hsd doesn't boot is to ask the user if they want to generate or restore a new wallet. That can be a little disturbing (out of scope for this PR).

If you follow the above procedure but with the patch in place, the rescan will initiate automatically but again, without the user really knowing whats going on.

And finally, if you just upgrade a non corrupt wallet, everything goes fine. You'll see this in the log:

```

[I:2020-08-12T18:47:32Z] (wallet) Loaded primary wallet (id=primary, wid=0, address=hs1qx4f6l5crrz3fzsqusqrwd6nmjl8kyllya037uq)
[W:2020-08-12T18:47:32Z] (wallet) Checking change address corruption...
[I:2020-08-12T18:47:32Z] (wallet) Checking wallet (id=primary, wid=0).
[I:2020-08-12T18:47:32Z] (wallet) All change addresses present.

```
